Sources/GRPC/ReadWriteStates.swift

```diff
@@ -0,0 +1,191 @@
+    var buffer = allocator.buffer(capacity: 0)
+    // TODO: add support for compression.
+    self.writer.write(data, into: &buffer, usingCompression: .none)
+
+    // If we only expect to write one message then we're no longer writable.
+    if case .one = self.expectedCount {
+      self.canWrite = false
+    }
+
+    return .success(buffer)
+  }
+}
+
+/// The read state of a stream.
+struct ReadState {
+  /// Whether the stream may read.
+  internal private(set) var canRead: Bool = true
+
+  /// The arity of the reading stream.
+  var expectedCount: MessageCount
+
+  /// A reader which accepts bytes in the gRPC wire-format and produces sequences of bytes which
+  /// may be decoded into protobuf `Message`s.
+  var reader: LengthPrefixedMessageReader
+
+  init(expectedCount: MessageCount, reader: LengthPrefixedMessageReader) {
+    self.expectedCount = expectedCount
+    self.reader = reader
+  }
+}
+
+enum MessageReadError: Error {
+  /// Too many messages were read.
+  case cardinalityViolation
+
+  /// Enough messages were read but bytes there are left-over bytes.
+  case leftOverBytes
+
+  /// Message deserialization failed.
+  case deserializationFailed
+
+  /// An invalid state was encountered. This is a serious implementation error.
+  case invalidState
+}
+
+extension ReadState {
+  /// Consume the given `buffer` then attempt to read and subsequently decode length-prefixed
+  /// serialzed messages.
+  ///
+  /// For `.unary` arity this function will produce **at most** 1 message. If a message has been
+  /// produced then subsequent calls will result in an error.
+  ///
+  /// - Parameter buffer: The buffer to read from.
+  /// - Parameter as: The type of `Message` to decode.
+  mutating func readMessage<T: Message>(
+    _ buffer: inout ByteBuffer,
+    as: T.Type = T.self
+  ) -> Result<[T], MessageReadError> {
+    guard self.canRead else {
+      return .failure(.cardinalityViolation)
+    }
+
+    self.reader.append(buffer: &buffer)
+    var messages: [T] = []
+
+    // Pull out as many messages from the reader as possible.
+    do {
+      while var serializedBytes = try? self.reader.nextMessage() {
+        // Force unwrapping is okay here: we will always be able to read `readableBytes`.
+        let serializedData = serializedBytes.readData(length: serializedBytes.readableBytes)!
+        messages.append(try T(serializedData: serializedData))
+      }
+    } catch {
+      return .failure(.deserializationFailed)
+    }
+
+    // If this a unary stream we need to validate the number of messages we decoded. Zero is fine
+    // becuase the payload may be split across frames.
+    switch (self.expectedCount, messages.count) {
+    case (.one, 1):
+      self.canRead = false
+      // We shouldn't have any bytes leftover after reading a single message.
+      if self.reader.hasBytes {
+        return .failure(.leftOverBytes)
+      }
+
+    case (.one, 2...):
+      return .failure(.cardinalityViolation)
```

Comment:
Should we also set `self.canRead = false`, or is this implied by returning `.failure`?

Reply:
A quick reply would be appreciated :-)

Reply:
It's implied by returning `.failure` but I think this should be more explicit. Same goes for `WriteState`.